### ISSUE TITLE
Orphan rules work differently when supertrait is specified in trait bounds

### ISSUE THREAD

**Issue by zakarumych** [rust-lang/rust#85371]
The following code is expected to compile without errors.

```rust
pub trait Foo {}
pub trait Bar<T>: Foo {}

pub trait Foo2 {}
pub trait Bar2<T>: Foo2 {}


impl<T> Foo2 for T where T: Foo {}
impl<U, T> Bar2<U> for T where T: Bar<U> {}

impl<T> Foo2 for Option<T> where T: Foo2 {}
impl<U, T> Bar2<U> for Option<T> where T: Bar2<U> {}
```

However possible conflict implementation in upstream crates is detected.

Rewriting 3rd impl as:
```rust
impl<U, T> Bar2<U> for T where T: Bar<U> + Foo {}
```
i.e. adding `Foo` bound,
solves the issue. Which should not be required as `Foo` is supertrait of `Bar`.

### Meta

`rustc --version --verbose`:
```
rustc 1.52.1 (9bc8c42bb 2021-05-09)
binary: rustc
commit-hash: 9bc8c42bb2f19e745a63f3445f1ac248fb015e53
commit-date: 2021-05-09
host: x86_64-unknown-linux-gnu
release: 1.52.1
LLVM version: 12.0.0
```

[playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8c75cb5cd282f0ebe3bf30b29dc773c5)

**Comment by FabianWolff** [comment]
Slightly reduced example:
```rust
pub trait Foo {}
pub trait Bar<T>: Foo {}
pub trait Baz<T> {}

impl<U, T> Baz<U> for T where T: Bar<U> {}
//impl<U, T> Baz<U> for T where T: Bar<U> + Foo {}

impl<U, T> Baz<U> for Option<T> where T: Baz<U> {}
```
```
error[E0119]: conflicting implementations of trait `Baz<_>` for type `std::option::Option<_>`:
 --> src/lib.rs:8:1
  |
5 | impl<U, T> Baz<U> for T where T: Bar<U> {}
  | --------------------------------------- first implementation here
...
8 | impl<U, T> Baz<U> for Option<T> where T: Baz<U> {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `std::option::Option<_>`
  |
  = note: downstream crates may implement trait `Bar<_>` for type `std::option::Option<_>`

error: aborting due to previous error
```
